hbase-server/src/main/java/org/apache/hadoop/hbase/master/ServerManager.java

```diff
@@ -324,8 +324,23 @@ public void regionServerReport(ServerName sn, ServerMetrics sl) throws YouAreDea
       // the ServerName to use. Here we presume a master has already done
       // that so we'll press on with whatever it gave us for ServerName.
       if (!checkAndRecordNewServer(sn, sl)) {
-        LOG.info("RegionServerReport ignored, could not record the server: " + sn);
-        return; // Not recorded, so no need to move on
+        // Master already registered server with same (host + port) and higher startcode.
+        // This can happen if regionserver report comes late from old server
+        // (possible race condition), by that time master has already processed SCP for that
+        // server and started accepting regionserver report from new server i.e. server with
+        // same (host + port) and higher startcode.
+        // The exception thrown here is not meant to tell the region server it is dead because if
+        // there is a new server on the same host port, the old server should have already been
+        // dead in ideal situation.
+        // The exception thrown here is to skip the later steps of the whole regionServerReport
+        // request processing. Usually, after recording it in ServerManager, we will call the
+        // related methods in AssignmentManager to record region states. If the region server
+        // is already dead, we should not do these steps anymore, so here we throw an exception
+        // to let the upper layer know that they should not continue processing anymore.
+        final String errorMsg = "RegionServerReport ignored, could not record the server: " + sn
+          + " . Consider yourself dead as server with higher startcode is already registered.";
```

Comment:
This doesn't need to block merging, but given the comments above, doesn't the message here seem odd? The message seems to be addressing the regionserver. But you say in the comment that this error is not sent to the regionserver and is only meant to signal upper layers to skip processing the dead server. 

I wonder if we shouldn't have a new exception type or at least clarify the message (which is not just thrush but also logged)

Reply:
The exception is sent to regionserver but the purpose is bit different. So not only we don't want to process the report but also we want to let the regionserver know that it should have been dead.
Let me see how I can make this clear in the comments.

Reply:
We can not send this message to regionserver as it is already dead... I have repeated these words many times, on the jira issue and also here...

Reply:
I meant to say that master sends it as if regionserver is still alive. But in reality it is dead by the time master receives the RPC call.

Reply:
I think the comment is clear. To me the log message needs to change

Reply:
How about
```
final String errorMsg = "RegionServerReport received from " + sn
  + ", but another server with the same name and higher startcode is already registered, ignoring";
```

Reply:
Yea, that sounds great

Reply:
Sure let me make changes today.